### ISSUE TITLE
Future of base image

### ISSUE THREAD

**Issue by domire8** [aica-technology/docker-images#94]
I wanted to discuss with you @LouisBrunner and @eeberhard how you imagine the future of our base images here. Currently, there is the problem that the base images are not versioned (related to #82) so if the CI rebuilds the `iron-devel` image for example, the old one is lost.

Now that we also build control libraries and modulo, is the idea that we just package them together similar to the backend-base and have a tagged ros2_control image that will most likely not change until the next version of ros?

**Comment by LouisBrunner** [comment]
In my mind, and as you said, I would see the images in this repository becoming pure ros2/ros2_control images. Control-libraries and Modulo would then be added later on a need-to basis (and in the final `base` image).

On the point of versioning, I would still version the images we produce but it doesn't mean we can't have a more generic tag as well.
For example: let's say we define the first Iron image as `1.0.0-iron` and `iron` and then increase the minor/patch version when we make changes, e.g. `1.1.0-iron` and `iron`. In the end we always have a latest `iron` but we can also pin exact builds. Once we upgrade to the next one we can just increase the major tag, e.g. `2.0.0-jjj`.
If the packages available or configuration setup changes in any way in between builds I think that would be better instead of a single rolling tag. In this case, we get both convenience and exactness.

On the point of `devel`, do we still need it if we don't have control-libraries bundled?

**Comment by domire8** [comment]
Okay, that would make sense I think.

We don't need the devel anymore no.

**Comment by eeberhard** [comment]
Yes, agreed with all points above. A generic tag for the base ros2_control image would be nice, followed by including the CL and Modulo dependencies as separately versioned images (thus removing those images from these workflows). Both of those are not blocking / high-priority but would certainly be nice to pick up in the next months when convenient.

**Comment by domire8** [comment]
I need a bit more clarification on one point:
Do we still pre-build a `ros2-modulo-control` image here (triggered manually with specified CL and modulo tags) for all our packages that depend on that or is there no image like that and each package needs to copy the dependencies in its own image? In other words, will for example the DSE use `ros2-modulo-control` as base or `ros2-control` and then copy the colcon workspace from `modulo`?

I would prefer option 1, but maybe that was already what you suggested.

**Comment by eeberhard** [comment]
The idea would be to build the output of `modulo` first and then copy the build output into the layers using `ros2-control` as base. This would make it much more flexible to change modulo version.

I imagine most dockerfiles would look something like:

```dockerfile
FROM ghcr.io/aica-technology/modulo:vX.Y.Z as modulo
FROM ghcr.io/aica-technology/ros2-control

COPY --from=modulo / /
```

**Comment by LouisBrunner** [comment]
@eeberhard In that scenario, `control-libraries` would be packaged with `modulo` no?
Isn't that contrary to what we have done so far as we avoid to bundle any packages together (apart in `backend-base`).

Wouldn't 
```Dockerfile
FROM ghcr.io/aica-technology/control-libraries:vX.Y.Z as control-libraries
FROM ghcr.io/aica-technology/modulo:vX.Y.Z as modulo
FROM ghcr.io/aica-technology/ros2-control

COPY --from=control-librarires / /
COPY --from=modulo /colcon ~/ws/install
```

be more flexible?

**Comment by eeberhard** [comment]
For brevity I only mentioned modulo, and my snippet was more pseudocode regarding the copy paths; I agree with you about control-libraries. In my earlier comment I wrote 

> [...] including the CL and Modulo dependencies as separately versioned images

Your snippet is much more clear

**Comment by LouisBrunner** [comment]
> > [...] including the CL and Modulo dependencies as separately versioned images

Nice, hadn't seen it
